src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -0,0 +1,1572 @@
+                var propertyMember = member as PropertyMemberAst;
+                if (propertyMember != null)
+                {
+                    add = propertyMember.IsStatic == isStatic;
+                }
+                else
+                {
+                    var functionMember = (FunctionMemberAst) member;
+                    add = functionMember.IsStatic == isStatic;
+                    foundConstructor |= functionMember.IsConstructor;
+                }
+
+                if (filterToCall != null && add)
+                {
+                    add = filterToCall(member);
+                }
+
+                if (add)
+                {
+                    results.Add(member);
+                }
+            }
+
+            //iterate through bases/interfaces
+            foreach (var baseType in typename.TypeDefinitionAst.BaseTypes)
+            {
+                var baseTypeName = baseType.TypeName as TypeName;
+                if (baseTypeName == null) continue;
+                var baseTypeDefinitionAst = baseTypeName._typeDefinitionAst;
+                results.AddRange(GetMembersByInferredType(new PSTypeName(baseTypeDefinitionAst), isStatic, filterToCall));
+            }
+
+            // Add stuff from our base class System.Object.
+            if (isStatic)
+            {
+                // Don't add base class constructors
+                if (filter == null)
+                {
+                    filterToCall = o => !IsConstructor(o);
+                }
+                else
+                {
+                    filterToCall = o => !IsConstructor(o) && filter(o);
+                }
+
+                if (!foundConstructor)
+                {
+                    results.Add(
+                        new CompilerGeneratedMemberFunctionAst(PositionUtilities.EmptyExtent, typename.TypeDefinitionAst,
+                            SpecialMemberFunctionType.DefaultConstructor));
+                }
+            }
+            else
+            {
+                // Reset the filter because the recursive call will add IsHidden back if necessary.
+                filterToCall = filter;
+            }
+            results.AddRange(GetMembersByInferredType(new PSTypeName(typeof(object)), isStatic, filterToCall));
+        }
+
+        private void AddMembersByInferredTypesClrType(PSTypeName typename, bool isStatic, Func<object, bool> filter, Func<object, bool> filterToCall,
+            List<object> results)
+        {
+            if (CurrentTypeDefinitionAst == null || CurrentTypeDefinitionAst.Type != typename.Type)
+            {
+                if (filterToCall == null)
+                {
+                    filterToCall = o => !IsMemberHidden(o);
+                }
+                else
+                {
+                    filterToCall = o => !IsMemberHidden(o) && filter(o);
+                }
+            }
+            IEnumerable<Type> elementTypes;
+            if (typename.Type.IsArray)
```

Comment:
Maybe it is the naming that is bad. The whole scenario seems to be to get the type from enumerables and arrays